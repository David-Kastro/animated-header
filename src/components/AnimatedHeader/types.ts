import { ViewStyle, TextStyle } from 'react-native';

export interface IHeaderProps {
  title: string;
  style?: ViewStyle;
  titleStyle?: TextStyle;
}

export interface ITitleProps {
  title: string;
  style?: TextStyle;
}

export interface IOptions {
  containerStyle?: ViewStyle;
  contentStyle?: ViewStyle;
  titleStyle?: TextStyle;
  headerTitleStyle?: TextStyle;
  headerStyle?: ViewStyle;
}

export type Animation = 'Fade' | 'Share';
