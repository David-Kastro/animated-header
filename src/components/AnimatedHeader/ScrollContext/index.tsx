import React, {useState, createContext, useContext} from 'react';
import {ScrollView as ScrollViewNative, ScrollViewProps} from 'react-native';

interface IProps {
  children: React.ReactNode;
}

export interface IScrollContext {
  opacity: number;
  maxOffset: number;
  offset: number;
  titleShowing: boolean;
  updateOffset(val: number): void;
}

const withinLimits = (val: number, min: number, max: number): number =>
  val > max ? max : val < min ? min : val;

export const ScrollContext = createContext<IScrollContext>({
  opacity: 0,
  maxOffset: 0,
  offset: 0,
  titleShowing: false,
  updateOffset: () => {},
});

export const useScroller = () => useContext(ScrollContext);

export const ScrollContextProvider = (props: IProps) => {
  const minOffset: number = 0;
  const maxOffset: number = 30;

  const [offset, setOffset] = useState(0);
  const [titleShowing, setTitleShowing] = useState(false);
  const [opacity, setOpacity] = useState(0);

  const updateOffset = (val: number) => {
    setOffset(withinLimits(val, minOffset, maxOffset));
    setTitleShowing(val > maxOffset);
    setOpacity(withinLimits((val * maxOffset) / 1000, 0, 1));
  };

  return (
    <ScrollContext.Provider
      value={{
        opacity: opacity,
        maxOffset: maxOffset,
        offset: offset,
        titleShowing: titleShowing,
        updateOffset: updateOffset,
      }}>
      {props.children}
    </ScrollContext.Provider>
  );
};

export const ScrollView = (props: IProps & ScrollViewProps) => {
  const {updateOffset} = useScroller();

  return (
    <ScrollViewNative
      {...props}
      onScroll={({nativeEvent}) => {
        updateOffset(nativeEvent.contentOffset.y);
      }}
      scrollEventThrottle={200}>
      {props.children}
    </ScrollViewNative>
  );
};

export default ScrollContextProvider;
