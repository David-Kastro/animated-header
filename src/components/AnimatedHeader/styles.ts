import {StyleSheet, TextStyle, ViewStyle} from 'react-native';

interface IStyles {
  flexView: ViewStyle;
  container: ViewStyle;
  header: ViewStyle;
  headerTitle: TextStyle;
  content: ViewStyle;
  title: TextStyle;
}

const styles: IStyles = {
  flexView: {
    flex: 1,
  },
  container: {
    backgroundColor: '#fff',
    flex: 1,
    paddingHorizontal: 10,
  },
  header: {
    justifyContent: 'center',
    alignItems: 'center',
    height: 45,
    shadowRadius: 0,
    backgroundColor: '#fff',
    shadowOffset: {
      width: 0,
      height: 1,
    },
    zIndex: 9,
  },
  headerTitle: {
    fontWeight: '700',
    color: '#000',
  },
  content: {
    marginTop: 20,
    flex: 1,
  },
  title: {
    marginTop: 10,
    fontSize: 28,
    fontWeight: '700',
    color: '#000',
  },
};

export default StyleSheet.create(styles);
