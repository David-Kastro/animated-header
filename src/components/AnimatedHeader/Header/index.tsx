import React, {useEffect, useState} from 'react';
import {View, Text, Animated, Easing, TextStyle, ViewStyle} from 'react-native';
import {useScroller} from '../ScrollContext';

import styles from '../styles';

interface IProps {
  title: string;
  options?: IOptions;
}

interface IOptions {
  headerStyle?: ViewStyle;
  headerTitleStyle?: TextStyle;
}

const AnimatedHeader = (props: IProps) => {
  const {title, options} = props;

  const {titleShowing, opacity} = useScroller();

  const [titleFade] = useState(new Animated.Value(0));

  useEffect(() => {
    titleShowing === false &&
      Animated.timing(titleFade, {
        toValue: 0,
        duration: 100,
        useNativeDriver: true,
        easing: Easing.sin,
      }).start();

    titleShowing === true &&
      Animated.timing(titleFade, {
        toValue: 1,
        duration: 100,
        useNativeDriver: true,
        easing: Easing.sin,
      }).start();
  });

  return (
    <View
      style={{
        ...(options?.headerStyle || styles.header),
        shadowOpacity: opacity,
        elevation: opacity * 6,
      }}>
      <Animated.View style={{opacity: titleFade}}>
        <Text style={options?.headerTitleStyle || styles.headerTitle}>
          {title}
        </Text>
      </Animated.View>
    </View>
  );
};

export default AnimatedHeader;
