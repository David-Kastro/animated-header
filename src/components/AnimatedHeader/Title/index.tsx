import React from 'react';
import {Text, Animated, TextStyle} from 'react-native';
import {useScroller} from '../ScrollContext';

import styles from './styles';

interface IProps {
  title: string;
  style?: TextStyle;
}

const Title = (props: IProps) => {
  const {title, style} = props;

  const {opacity} = useScroller();

  return (
    <Animated.View style={{opacity: 1 - opacity}}>
      <Text style={style || styles.title}>{title}</Text>
    </Animated.View>
  );
};

export default Title;
