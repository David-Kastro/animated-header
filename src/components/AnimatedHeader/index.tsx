import React from 'react';
import {Text, View, SafeAreaView, TextStyle, ViewStyle} from 'react-native';
import Header from './Header';
import {ScrollContextProvider, ScrollView} from './ScrollContext';

import styles from './styles';

interface IProps {
  title: string;
  animation?: String;
  backAction?: () => void;
  icon?: String;
  options?: IOptions;
  children?: React.ReactNode;
}

interface IOptions {
  containerStyle?: ViewStyle;
  contentStyle?: ViewStyle;
  titleStyle?: TextStyle;
}

const AnimatedHeader = (props: IProps) => {
  const {title, options, children} = props;

  return (
    <ScrollContextProvider>
      <SafeAreaView style={styles.flexView}>
        <Header title={title} />
        <ScrollView
          contentInsetAdjustmentBehavior="automatic"
          style={options?.containerStyle || styles.container}>
          <Text style={options?.titleStyle || styles.title}>{title}</Text>
          <View style={options?.contentStyle || styles.content}>
            {children}
          </View>
        </ScrollView>
      </SafeAreaView>
    </ScrollContextProvider>
  );
};

export default AnimatedHeader;
