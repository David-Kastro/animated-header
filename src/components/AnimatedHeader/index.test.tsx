import React from 'react';
import { shallow } from 'enzyme';
import AnimatedHeader from './';

describe('<AnimatedHeader>', () => {
  it('renders the component', () => {
    const component = shallow(<AnimatedHeader title="Title" />);
    expect(component).toMatchSnapshot();
  });
});
