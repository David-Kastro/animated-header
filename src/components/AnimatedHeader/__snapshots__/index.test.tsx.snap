// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AnimatedHeader> renders the component 1`] = `
<ScrollContextProvider>
  <ForwardRef(SafeAreaView)
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <AnimatedHeader
      title="Title"
    />
    <ScrollView
      contentInsetAdjustmentBehavior="automatic"
      style={
        Object {
          "backgroundColor": "#fff",
          "flex": 1,
          "paddingHorizontal": 10,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#000",
            "fontSize": 28,
            "fontWeight": "700",
            "marginTop": 10,
          }
        }
      >
        Title
      </Text>
      <View
        style={
          Object {
            "flex": 1,
            "marginTop": 20,
          }
        }
      />
    </ScrollView>
  </ForwardRef(SafeAreaView)>
</ScrollContextProvider>
`;
